### PR TITLE
DEVX-1142 Adds a new demo for Replicator on k8s to Confluent Cloud

### PR DESCRIPTION
**Pull request by rspurgeon**
A new demonstration for CP Operator utilizing CP Replicator to replicate data to a Confluent Cloud cluster.

Staging Docs:
http://staging-docs.confluent.io.s3-website-us-west-2.amazonaws.com/PR/2996/5.3.1/tutorials/examples/kubernetes/replicator-gke-cc/docs/index.html

### CODE REVIEW

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Grammar: `This demonstrations features a deployment`

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Technically speaking, there is something called an `environment` in CCloud.  So maybe better to just say cluster?

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Consider adding to the docs explanation on how to destroy resources.  Perhaps I'm missing it, but I don't see that on this page.

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Add hyperlink to the repo

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst:65]: need to checkout a specific branch?

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: If this is repeated between gke-base and replicator-gke-cc, then make an `includes` directory and use the include directive, so it's only one copy.

**Review by ybyzek — COMMENTED** [review]
@rspurgeon cool!  I left round 1 of comments

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: Yeah, I intended to use the term environment here to mean "ccloud environment".

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: WIP

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst:65]: Yeah, as mentioned in DM, this won't work until merged

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: I'll add the term cluster to the components list here

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Are these required?  Do they need to be posted upfront?  Or can they be moved to advanced, e.g. https://docs.confluent.io/current/tutorials/examples/kubernetes/gke-base/docs/index.html#advanced-demo-usage

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst:33]: Somewhere add link to gke-base

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: 1. `You can obtain these...` -- can you find a doc link or include a screenshot with this info?
2. BTW, this can also be done via the CCloud CLI.

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Should these overrides go into the Advanced section?

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst:69]: If a user ran gke-base, is it an incremental run for Replicator/CCloud?  Or they need to tear down gke-base and then provision/run this demo?

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: 2. @ybyzek Can you retrieve the secret on the CLI?  All the documentation and CLI commands look to me like you can't grab that

Comment (ybyzek) [reply kubernetes/replicator-gke-cc/docs/index.rst]: @rspurgeon IIUC you can't retrieve a key's secret from an existing key/secret pair, neither through the CLI nor the GUI.  However, I brought up the CLI as a means to create a new key/secret pair (which can also be done through the GUI).

**Comment by rspurgeon** [comment]
All sections of documentation have some content now.  Staging docs have been built.  @ybyzek if you'd like to give this another spin and review the docs I'd appreciate it.

cc @gAmUssA @markteehan

**Comment by ybyzek** [comment]
FYI copying docs team for review.

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Two warnings back to back diminish the second warning.   Consider changing this to `note` instead of `warning`, or add it into the chart above

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Demo uses real CCloud resources too :)

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]:
```suggestion
Clone the `Confluent examples repository <https://github.com/confluentinc/examples>`__ and change directories on your terminal into the ``kubernetes/replicator-gke-cc`` directory.
```

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Consider whether to move this under "GKE Setup"

File: kubernetes/gke-base/docs/index.rst
Comment (ybyzek) [inline kubernetes/gke-base/docs/index.rst:63]: Aren't a few of these paragraphs shared with the previous demo?  Assuming yes, consider making an `include` and pulling it in so to not duplicate

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: `|ccloud|`

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst:162]: ...and in the sample output below, the Bootstrap Server value is foob-bar...

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Is there a reason to capitalize Bootstrap Server?

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Everything from "Confluent Cloud Account Setup" to "API Key and Secret Configuration" is super useful and important for first time users, but it's not required for every run.  Consider creating a new section for "First-time setup" (or something like that) and move it there, then refer to it.  Just to separate it from "Running the Demo" which would be required each and every run.

Also cc: @joel-hamill because IMO what @rspurgeon wrote up could be integrated into the quickstart or as an include somewhere.

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: This is already part of "Running the demo", so it's redundant to have "Run".  That's why "First time run" should precede perhaps

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst:316]: You haven't introduced `stock-trades` yet, so this feels a bit unexpected.  Consider giving an overview earlier in the doc page about the architecture/dataflow

Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst:316]: Add some explanation to this page.  Most people are viewing this for the first time and have no idea what they should be gleaning from it.

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Explain which cluster is which, and why broker stats are available on one but not the other.

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: The word `Controller` is repeated

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Forward slash instead of back slash?

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst:392]:
```suggestion
Deploying Confluent Replicator with Confluent Operator
```

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst]: Add a reference to https://docs.confluent.io/current/tutorials/examples/kubernetes/gke-base/docs/index.html#connector-deployments

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [inline kubernetes/replicator-gke-cc/docs/index.rst:313]: Consumer lag --> I think this should go into Highlights, not Validation.  (The presence of stock-trades in CCloud is validation enough)

File: kubernetes/gke-base/docs/index.rst
Comment (ybyzek) [inline kubernetes/gke-base/docs/index.rst:207]: Include a screenshot of C3 showing the Connect page, you should see the Replicator connector running

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: yes, nice catch, ty... improved the entire warning

File: kubernetes/gke-base/docs/index.rst
Comment (rspurgeon) [reply kubernetes/gke-base/docs/index.rst:63]: The syntax is included from a shared file but I prefer the surrounding text to be explict in the doc.

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst:392]: My intention with this Highlight is to show how any connector could be deployed using this method where the Replicator config is the example.  IMO it adds further value to the highlight, although I will consider expanding the language to specificy Replicator

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: I evolved the idea across these demos from standard K8s mainfests to a Helm chart.  I'm worried it could get confusing as they are comparable, but the Helm version here deploying replicator allows for better varability

**Comment by joel-hamill** [comment]
ping @londoncalling the current on-call for docs

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst:313]: Please clarify "consumer lag -->" in this comment

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst:162]: Maybe move your comment to the foob-bar line to show me what you're seeing?  Looks proper on my screen

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: Because this highlights Replicated environments my intention was that the demo be ran through the entirety each time.  If the user runs the demo over and over without the cluster creation piece, they will end up with non-identical topics.   I'm surely not articulating that well with the exception of the Troubleshooting section at the bottom

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: As mentioned in the above comment, i'm not really intending for there to be a "first run"

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: nope... resolving thank you

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: Agree that the heading was redundant, removed

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [reply kubernetes/replicator-gke-cc/docs/index.rst:313]: Clarification: referring to the C3 screenshot of monitoring Replicator

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst:316]: What do you suggest more than the existing "This will give you a view of the Confluent Replicator consumer group as it replicates the stock-trades topics to Confluent Cloud." ?  Should we just link to C3 docs instead of reproducing a C3 primer here?

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [reply kubernetes/replicator-gke-cc/docs/index.rst]: Gotcha.  Instead, how about calling out the datagen piece explicitly here to show how it's configured in this phase 2 part of the demo?  So highlights would have two things side by side: datagen connector producing to GKE and Replicator copying to CCloud?

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (ybyzek) [reply kubernetes/replicator-gke-cc/docs/index.rst:316]: Just need some lightweight observations about what can be seen here, besides just "Monitor."  i.e., this shows monitoring what metrics exactly?

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: in this case the stock-trades data gen is configured in the same manner as the clicks datagen, using the demo phase 1 method, so my concern applies about mixing the methods to the reader.

If you want a separate highlight for datagen I can add that, but I don't want to focus on how it's deployed vs how the Replicator config is deployed, as they are similar but different.

If i could start over, the demos would all use helm charts instead of the phase 1 method

Comment (ybyzek) [reply kubernetes/replicator-gke-cc/docs/index.rst]: > If you want a separate highlight for datagen I can add that, but I don't want to focus on how it's deployed vs how the Replicator config is deployed, as they are similar but different.

Consider the user who stumbles upon this docs page, they will want to know how to deploy the datagen connector, so this page needs to add that content.  How much you compare/contrast that to how Replicator is deployed is up to you.

But I am concerned about this statement: "If i could start over, the demos would all use helm charts instead of the phase 1 method". 

1. I don't understand (a) what are the two methods and (b) why one is better than the other -- can you please elaborate with an example?

2. Based on above, the next decision is whether to "fix" it sooner or later.  For now, at least please file a jira so we track internally.

**Review by londoncalling — CHANGES_REQUESTED** [review]
@rspurgeon please do not merge until I have tested this, back to you soon

**Comment by rspurgeon** [comment]
> @rspurgeon please do not merge until I have tested this, back to you soon

No problem @londoncalling ... I'm also running the doc build locally as I iterate on it, I'll also rebuild the staging docs as I go.

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst:316]: The stock-trades datagen is mentioned briefly in the Demo components above

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: I'll file a Jira on the potential improvements to the demo workflow, but the user looking for answers to deploying datagen connectors will need to find that elsewhere (DEVX-1243)

File: kubernetes/replicator-gke-cc/docs/index.rst
Comment (rspurgeon) [reply kubernetes/replicator-gke-cc/docs/index.rst]: Added a note to this section to indicate this intention

File: kubernetes/gke-base/docs/index.rst
Comment (rspurgeon) [reply kubernetes/gke-base/docs/index.rst:207]: This isn't supported right now w/ standard Operator deployment, so i'm going to kick this can for now.  Jira filed (DEVX-1244)

**Comment by londoncalling** [comment]
> > @rspurgeon please do not merge until I have tested this, back to you soon
> 
> No problem @londoncalling ... I'm also running the doc build locally as I iterate on it, I'll also rebuild the staging docs as I go.

@rspurgeon , cool, I tested this with our docs builds. When I build this locally  I get the following docs build error:

```
Warning, treated as error:
/Users/vicky/projects/docs/tutorials/examples/kubernetes/replicator-gke-cc/docs/index.rst:: WARNING: document isn't included in any toctree
```
**How to Fix**
To fix as a part of this PR, please add the following line to the hidden `toctree` files listing at the bottom of `tutorials/index.rst`:

```
 examples/kubernetes/replicator-gke-cc/docs/index
```
**Example**
For example, In my local tests, I added it after `examples/kubernetes/gke-base/docs/index` as shown here:

```
.. toctree::
    :maxdepth: 1
    :hidden:

    cp-demo/docs/index
    examples/ccloud/docs/index
    examples/microservices-orders/docs/index
    examples/kubernetes/gke-base/docs/index
    examples/kubernetes/replicator-gke-cc/docs/index
    examples/replicator-schema-translation/docs/index
    event_driven_apps_tutorial
```

**Review by londoncalling — APPROVED** [review]
@rspurgeon approving since I think you have a handle on this

**Review by ybyzek — APPROVED** [review]
@rspurgeon awesome demo, really showcases capabilities across the platform and cloud.  As we discussed separately, let's merge the PR as it is to make it more readily available, and we'll iterate over it in separate PRs.

**Comment by rspurgeon** [comment]
> @rspurgeon awesome demo, really showcases capabilities across the platform and cloud. As we discussed separately, let's merge the PR as it is to make it more readily available, and we'll iterate over it in separate PRs.

Will be merging today.  Thanks @ybyzek and @londoncalling for the reviews

**Pull request merged**
